### PR TITLE
Feature/session module

### PR DESCRIPTION
**Pull request by laurentiuonac**
Created a base implementation for a session module. Current main purposes are:

- to hold data about login session of a specific user 
- provide a way to listen for login state changes
- offer the ability to configure network layer for authenticated API calls

Can be expanded to hold other elements that are tied to the session lifecycle (like user preferences, settings, etc).

### CODE REVIEW

File: session/src/main/java/com/tapptitude/session/SessionManagerImpl.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import com.tapptitude.session.storage.LocalSessionDataManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+import org.koin.core.component.KoinComponent
+import org.koin.core.component.inject
+
+internal class SessionManagerImpl : SessionManager, KoinComponent {
+    private val localSessionDataManager: LocalSessionDataManager by inject()
```
Comment (cristidregan) [inline session/src/main/java/com/tapptitude/session/SessionManagerImpl.kt:15]: Wouldn't it be better to have localSessionDataManager as a contructor injection?

File: session/src/main/java/com/tapptitude/session/model/LoginState.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package com.tapptitude.session.model
+
+sealed class LoginState {
```
Comment (cristidregan) [inline session/src/main/java/com/tapptitude/session/model/LoginState.kt:3]: I think these brackets are unnecessary.

File: session/src/main/java/com/tapptitude/session/model/LoginState.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+}
+
+class LoggedIn(
+    val sessionToken: String,
+    val userId: String
+) : LoginState() {
+}
```
Comment (cristidregan) [inline session/src/main/java/com/tapptitude/session/model/LoginState.kt:10]: I think these brackets are unnecessary.

**Review by cristidregan — APPROVED** [review]
Looks great!

File: session/src/main/java/com/tapptitude/session/SessionManagerImpl.kt (hunk 1 above)
Comment (laurentiuonac) [reply session/src/main/java/com/tapptitude/session/SessionManagerImpl.kt:15]: Great catch ;)

File: featureHome/src/main/java/com/tapptitude/featurehome/presentation/HomeViewModel.kt [hunk 4]
```diff
@@ -23,4 +37,33 @@ class HomeViewModel(private val loadImageUseCase: LoadImageUseCase) : ViewModel(
...
+    }
+
+    private fun doLogin() {
+        val randomId = (System.currentTimeMillis() % 10000).toString()
+        sessionManager.onLoggedIn("SampleToken", "userId$randomId")
+    }
+
+    private fun doLogout() {
+        val loggedInState = sessionManager.currentLoginStateFlow.value as? LoggedIn
+
```
Comment (alexhasmasan) [inline featureHome/src/main/java/com/tapptitude/featurehome/presentation/HomeViewModel.kt:58]: I think the extra empty line could be removed

**Review by alexhasmasan — APPROVED** [review]

**Review by andreealaura16 — APPROVED** [review]

**Pull request merged**